docs/writing-controllers.md

```diff
@@ -0,0 +1,1489 @@
+
+An event is not an action, so the identifier for an event should not describe something that the controller needs to do, but rather something that has has happened to the controller.
+
+You may find it helpful to pretend that the word "on" precedes the identifier (but do not include this in the identifier itself).
+
+🚫 **`receiveMessage` sounds like a command**
+
+```typescript
+const name = 'FooController';
+
+export type FooControllerReceiveMessageEvent = {
+  type: `${typeof name}:receiveMessage`;
+  payload: () => void;
+};
+```
+
+🚫 **The event identifier includes "on"**
+
+```typescript
+const name = 'FooController';
+
+export type FooControllerOnMessageReceivedEvent = {
+  type: `${typeof name}:onMessageReceived`;
+  payload: () => void;
+};
+```
+
+✅ **The event identifier sounds like a statement and does not include "on"**
+
+```typescript
+const name = 'FooController';
+
+export type FooControllerMessageReceivedEvent = {
+  type: `${typeof name}:messageReceived`;
+  payload: () => void;
+};
+```
+
+## Use standard naming for event types
+
+Types for messenger events should follow this naming:
+
+```
+Controller name + Event identifier without controller name prefix + "Event"
+```
+
+🚫 **Type name does not include controller name and does not end with "Action"**
+
+```typescript
+const name = 'FooController';
+
+export type MessageReceived = {
+  type: `${typeof name}:messageReceived`;
+  payload: () => void;
+};
+```
+
+✅ **Type name now matches controller name + event identifier + "Event"**
+
+```typescript
+const name = 'FooController';
+
+export type FooControllerMessageReceivedEvent = {
+  type: `${typeof name}:messageReceived`;
+  payload: () => void;
+};
+```
+
+## Define and export a type union for internal action types
+
+A controller should define and export a type union that holds all of its actions. This type union makes it easy to define the type for the controller's messenger.
+
+The name of this type should be `${ControllerName}Actions`.
+
+This type should be only passed to `RestrictedControllerMessenger` as the 2nd type parameter. It should _not_ be included in its 4th type parameter, as that is is used for external events.
+
+🚫 **`FooController['type']` is passed as the 4th type parameter**
+
+```typescript
+export type FooControllerActions =
+  | FooControllerUpdateCurrencyAction
+  | FooControllerUpdateRatesAction;
+
+export type FooControllerMessenger = RestrictedControllerMessenger<
+  'FooController',
+  FooControllerActions,
+  never,
+  FooControllerActions['type'],
+  never
+>;
+```
+
+✅ **`never` is passed as the 5th type parameter (assuming no external events)**
+
+```typescript
+export type FooControllerActions =
+  | FooControllerUpdateCurrencyAction
+  | FooControllerUpdateRatesAction;
+
+export type FooControllerMessenger = RestrictedControllerMessenger<
+  'FooController',
+  FooControllerActions,
+  never,
+  never,
+  never
+>;
+```
+
+## Define and export internal event types as one
```

Comment:
For consistency with https://github.com/MetaMask/core/pull/4110#discussion_r1670856411

```suggestion
## Define and export a type union for internal event types
```